File: snippets/lua/nui-callback.lua
-- CFX NUI Callback Pattern
-- Register a callback that the NUI (browser) can invoke via fetch POST.
-- Always call cb() to complete the request.

RegisterNUICallback('actionName', function(data, cb)
    -- data: JSON body from the NUI fetch request
    -- cb: callback function, must be called to resolve the fetch

    print('NUI sent: ' .. tostring(data.someField))

    -- Do your logic here

    cb('ok') -- or cb({ success = true, result = someValue })
end)

-- To open NUI and give it focus:
-- SendNUIMessage({ type = 'open', payload = {} })
-- SetNuiFocus(true, true)

-- To close NUI:
-- SendNUIMessage({ type = 'close' })
-- SetNuiFocus(false, false)

File: snippets/javascript/nui-callback.js
// CFX NUI Callback Pattern (JavaScript - Client Script Side)
// Register a NUI callback in a JS client script and send messages to the NUI browser.
// This is the CFX-client-side counterpart to the browser's window.addEventListener('message').

// Send a message from client script to NUI browser
// The NUI browser receives this via window.addEventListener('message', ...)
function openMenu(data) {
    SendNUIMessage({ type: 'open', payload: data });
    SetNuiFocus(true, true);
}

// Register a callback that the NUI browser can invoke via fetch POST
// Browser calls: fetch(`https://${GetParentResourceName()}/closeMenu`, { method: 'POST', body: JSON.stringify({}) })
RegisterNuiCallbackType('closeMenu');
on('__cfx_nui:closeMenu', (data, cb) => {
    SetNuiFocus(false, false);
    cb({ ok: true });
});

// Clean up on resource stop
on('onResourceStop', (resourceName) => {
    if (GetCurrentResourceName() !== resourceName) return;
    SetNuiFocus(false, false);
});
